src/reducer/players-reducer.js

```diff
@@ -93,6 +93,15 @@ const playersReducer = createReducer({
             players
         };
     },
+    [ADD_PUNITIVE_CARD]: (state, data) => {
+        const {name} = data;
+        const {players} = state;
+        const card = state.stockCards.splice(1, 1);
+        players[name].cards.push(card[0]);
+        return {
+            ...state
+        };
+    },
```

Comment:
splice et push modifie les tableaux (il faut éviter surtout dans le reducer sinon redux ne détecte pas les modifications correctement) :
```
[ADD_PUNITIVE_CARD]: (state, {playerName}) => ({
        ...state,
       stockCards: state.stockCards.slice(1),
       players: {
            ...state.players,
           [playerName]: {
                ...state.players[playerName],
               cards: state.players[playerName].cards.concat(state.stockCards[0])
           }
       }
})
```
Il doit y avoir une façon plus élégante de faire via Ramda et  mergeDeepRight celà dit (pas besoin de répéter les ...state à chaque fois, il faut juste préciser les donner qui changent) : https://ramdajs.com/docs/#mergeDeepRight

Reply:
Le fait que l'on return tout le state du coup c'est pas pris en compte quand même ?

Reply:
En fait react-redux fait une vérification sur les références des objets pour savoir si le state a été modifié. Du coup ça risque de ne pas être pris en compte.
Et aussi, on ne peut plus revenir en arrière et faire du time-travel debugging :

https://stackoverflow.com/questions/39343700/why-immutable-state-with-redux

https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns#inserting-and-removing-items-in-arrays